src/libspark/keys.cpp

```diff
@@ -12,6 +13,19 @@ SpendKey::SpendKey(const Params* params) {
 	this->r.randomize();
 }
 
+SpendKey::SpendKey(const Params* params, const Scalar& r_) {
+    this->params = params;
+    this->r = r_;
+    std::vector<unsigned char> data;
+    r.serialize(data.data());
+    std::vector<unsigned char> result(CSHA512().OUTPUT_SIZE);
+    CHash512 hash512;
+    hash512.Write(data.data(), data.size()).Finalize(&result[0]);
```

Comment:
Recommend using domain separation for this hash instantiation to mitigate any risk of collision.